packages/flutter/test/material/dropdown_test.dart

```diff
@@ -3165,4 +3165,164 @@ void main() {
         ..rrect(rrect: const RRect.fromLTRBXY(0.0, 0.0, 112.0, 47.0, 2.0, 2.0), color: Colors.grey[50], hasMaskFilter: false)
     );
   });
+
+  testWidgets('Tapping a disabled item should not close DropdownButton', (WidgetTester tester) async {
+    String? value = 'first';
+
+    await tester.pumpWidget(
+      MaterialApp(
+        home: Scaffold(
+          body: StatefulBuilder(
+            builder: (BuildContext context, StateSetter setState) => DropdownButton<String>(
+              value: value,
+              items: const <DropdownMenuItem<String>>[
+                DropdownMenuItem<String>(
+                  enabled: false,
+                  child: Text('disabled'),
+                ),
+                DropdownMenuItem<String>(
+                  value: 'first',
+                  child: Text('first'),
+                ),
+                DropdownMenuItem<String>(
+                  value: 'second',
+                  child: Text('second'),
+                ),
+              ],
+              onChanged: (String? newValue) {
+                setState(() {
+                  value = newValue;
+                });
+              },
+            ),
+          ),
+        ),
+      ),
+    );
+
+    // Open dropdown.
+    await tester.tap(find.text('first').hitTestable());
+    await tester.pumpAndSettle();
+
+    // Tap on a disabled item.
+    await tester.tap(find.text('disabled').hitTestable());
+    await tester.pumpAndSettle();
+
+    // The dropdown should still be open, i.e., there should be one widget with 'second' text.
+    expect(find.text('second').hitTestable(), findsOneWidget);
+  });
+
+  testWidgets('Disabled item should not be focusable', (WidgetTester tester) async {
+    await tester.pumpWidget(
+      MaterialApp(
+        home: Scaffold(
+          body: DropdownButton<String>(
+            value: 'enabled',
+            onChanged: onChanged,
+            items: const <DropdownMenuItem<String>>[
+              DropdownMenuItem<String>(
+                enabled: false,
+                child: Text('disabled'),
+              ),
+              DropdownMenuItem<String>(
+                value: 'enabled',
+                child: Text('enabled'),
+              )
+            ],
+          ),
+        ),
+      ),
+    );
+
+    // Open dropdown.
+    await tester.tap(find.text('enabled').hitTestable());
+    await tester.pumpAndSettle();
+
+    try {
+      // try to get the focus node for disabled item
+      final FocusNode node = Focus.of(tester.element(find.text('disabled').hitTestable()));
+      // For disabled item, [Focus.of] should throw an exception,
+      // but it doesn't when platform=chrome and web-renderer=html when flutter test is run
+      // In that case, the node should not be able to request focus
+      expect(node.canRequestFocus, false, reason: 'Disabled menu item should not be able to request focus');
```

Comment:
Interesting -- so it does not throw an error only for web, but the `FocusNode` itself is set to not being able to request focus? Do you happen to know the reason this is the case?

cc/ @gspencergoog for comment, since he's very familiar with focus-related stuff!

Reply:
Yes, precisely.
I have tried to understand the reason to the problem, but it was difficult to understand all the underlying modules, being a beginner. I have asked out for it in the flutter discord and searched over in the flutter docs but have not got anything yet.

Reply:
`Focus.of` retrieves the focus node for the nearest `Focus` widget, and `FocusScope.of` retrieves the focus node for the nearest `FocusScope` widget. The `Focus.of` can also request either one by supplying `scopeOk: true`.

Also, you can test for existence of such a node without having to catch the assert, by just using `Focus.maybeOf` or `FocusScope.maybeOf` instead, and checking for null.

So, the test for this should look like:

```dart
Element disabled = tester.element(find.text('disabled').hitTestable()));
expect(Focus.maybeOf(disabled, scopeOk: true).canRequestFocus, false, reason: 'Disabled menu item should not be able to request focus');
```

I'm not sure why the web implementation might be failing, but you can dump the focus tree there with `debugDumpFocusTree();` to see what the hierarchy looks like. If that still doesn't give you enough information, you can turn on focus debugging by setting `_kDebugFocus = true;` at the top of [`focus_manager.dart`](https://cs.opensource.google/flutter/flutter/+/master:packages/flutter/lib/src/widgets/focus_manager.dart;l=20;drc=f82046b150ad9d54962ad8098df2babdad34c596).

Reply:
I understood the problem from the focus tree. There is a focus node in the disabled item for web because of web shortcuts. Should I remove the web shortcuts for disabled item?

```dart
if (kIsWeb) {
  child = Shortcuts(
  shortcuts: _webShortcuts,
  child: child,
  );
}
```

Also, the web shortcuts function says the following ([link](https://github.com/flutter/flutter/blob/f0934fdee35613e4eaf576f460d2a3ff75a1962f/packages/flutter/lib/src/material/dropdown.dart#L150))
```dart
// On the web, up/down don't change focus, *except* in a <select>
// element, which is what a dropdown emulates.
```

<details>
<summary>web implementation focus tree</summary>

```dart
Child 2: FocusScopeNode#66697(_ModalScopeState<dynamic> Focus Scope [IN FOCUS PATH])
│ context: FocusScope
│ IN FOCUS PATH
│ focusedChildren: FocusNode#38cd3([PRIMARY FOCUS])
│
├─Child 1: FocusNode#ba582(Shortcuts) // disabled item
│   context: Focus
│   NOT FOCUSABLE
│
└─Child 2: FocusNode#3ec9f(Shortcuts [IN FOCUS PATH])
    │ context: Focus
    │ NOT FOCUSABLE
    │ IN FOCUS PATH
    │
    └─Child 1: FocusNode#38cd3([PRIMARY FOCUS]) // enabled item
           context: Focus
           PRIMARY FOCUS
```
</details>
<details>
<summary>non web implementation focus tree</summary>

```dart
Child 2: FocusScopeNode#5c8f2(_ModalScopeState<dynamic> Focus Scope [IN FOCUS PATH])
│ context: FocusScope
│ IN FOCUS PATH
│ focusedChildren: FocusNode#7972a([PRIMARY FOCUS])
│
└─Child 1: FocusNode#7972a([PRIMARY FOCUS]) // enabled item
        context: Focus
        PRIMARY FOCUS
```
</details>

Reply:
Yes, it seems reasonable to not include the shortcuts for a disabled item, as long as you can determine that it really is disabled at the point where the shortcuts are inserted.

Reply:
I have added a constructor argument to ask from the user if it is enabled or not, if it's what you are asking for. If `enabled` is `false`, it is disabled.